### PR TITLE
OSD based parameter settings

### PR DESCRIPTION
**Pull request by vierfuffzig**
this is intended as a basic proposal for param osd documentation. ping @andyp1per

### CODE REVIEW

**Comment by andyp1per** [comment]
Looks good. I'll let @Hwurzburg fix up the grammar to his liking. 
For the mnemonics, this is the list of possible values:
```
"NONE"
"SERIAL_PROTOCOL"
"SERVO_FUNCTION"
"AUX_FUNCTION"
"FLIGHT_MODE"
"FAILSAFE_ACTION"
"FAILSAFE_ACTION_1"
"FAILSAFE_ACTION_2"
```
The failsafe action options mean different things on plane and copter which we should call out

**Comment by vierfuffzig** [comment]
updated respectively. would highly appreciate henry's review. ping @Hwurzburg

**Comment by Hwurzburg** [comment]
on my list...trying to get some flying in this weekend...

File: common/source/docs/common-paramosd.rst [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+   This feature my have limitations on certain parameters with exceptionally wide value ranges and adaptive incrementation steps. Selecting a 
+   different parameter using scroll selection will reset the type mnemonics to "none".
+
+
+Populating a screen's parameter slots using maxproxy's OSD module
+=================================================================
+ 
+OSD based parameter setting includes a `mavproxy module <https://ardupilot.org/mavproxy/>`__ to populate each screen's parameter slots.
+Connect your flight controller and load the OSD module by typing :code:`module load osd`
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:84]: Connect your flight controller, connect to MAVProxy, and....

File: common/source/docs/common-paramosd.rst [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+When one of the parameter screens is selected, adjustments can be made in disarmed state only. If a parameter screen is 
+selected inflight, the screen allows display only. This allows live monitoring of automatic parameter adjustments, 
+i.e. when performing autotune procedures or automatic servo trims.
+
+
+Menu navigation
+===============
+
+Use the RC transmitter's roll and pitch input to navigate through the parameter screen. UP / DOWN and LEFT / RIGHT will navigate through 
+the screen's parameter slots. The selected parameter will blink. Move right on the slected parameter to adjust its value using
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:36]: selected

File: common/source/docs/common-paramosd.rst [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+
+Additionally there's a "save and reboot" button on each screen that allows to save the adjusted settings and reboot the 
+flightcontroller if required.
+
+Accessing parameter screens
+===========================
+
+To enable the parameter screens, set :ref:`OSD5_ENABLE <OSD5_ENABLE>` and :ref:`OSD6_ENABLE <OSD6_ENABLE>` = 1.
+Paramater screens can be accessed in the same way the conventional osd screens are accessed.
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:25]: Parameter

File: common/source/docs/common-paramosd.rst [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+========================
+OSD based parameter menu
+========================
+
+.. image:: ../../../images/paramscreen.jpg
+    :target: ../_images/paramscreen.jpg
+    :width: 450px
+
+This allows to set and adjust paramaters using the ardupilot maxchip-based onboard OSD and an RC transmitter's stick input.
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:11]: parameter

Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst:11]: ArduPilot
Maxim chip based

File: common/source/docs/common-paramosd.rst [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+   OSD based parameter adjustment is designed to allow basic adjustments without any aditional data link or GCS setup required. However, 
+   some paramaters with wide value ranges might not be ideal for adjustment using parameter OSD.
+
+
+Parameter value and range mnemonics
+===================================
+
+OSD based parameter screens use mnemonics to account for complex functional parameter values and value ranges. This allows estimate useful 
+incrementation steps for individual paramater value ranges, as well as to display human-readable abbreviations for:
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:62]: parameter

File: common/source/docs/common-paramosd.rst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+
+- failsave actions
+
+.. note::
+
+   This feature my have limitations on certain parameters with exceptionally wide value ranges and adaptive incrementation steps. Selecting a 
+   different parameter using scroll selection will reset the type mnemonics to "none".
+
+
+Populating a screen's parameter slots using maxproxy's OSD module
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:80]: MAVProxy's

File: common/source/docs/common-paramosd.rst [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+.. note::
+
+   OSD based parameter adjustment is designed to allow basic adjustments without any aditional data link or GCS setup required. However, 
+   some paramaters with wide value ranges might not be ideal for adjustment using parameter OSD.
+
+
+Parameter value and range mnemonics
+===================================
+
+OSD based parameter screens use mnemonics to account for complex functional parameter values and value ranges. This allows estimate useful
```
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst:61]: Really confusing about mnemonics....what happens as you scroll thru params interactively in slot...what is displayed? are there presets for incr/min/max form the param metadata? are these used to provide reasonable values for those..but they have to be preloaded, right?

File: common/source/docs/common-paramosd.rst [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+.. image:: ../../../images/paramscreen.jpg
+    :target: ../_images/paramscreen.jpg
+    :width: 450px
+
+This allows to set and adjust paramaters using the ardupilot maxchip-based onboard OSD and an RC transmitter's stick input. 
+There's two additional OSD screens available (OSD5 and OSD6) that can hold up to 9 parameters each. Parameter values can be 
+adjusted in disarmed state only. The RC transmitter's roll and pitch stick input is used to navigate through the screen, pick 
+a parameter and adjust the parameter value.
+
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:15]: I think a few words about the concept of slots...that each screen has 9 slots, preselected via MAVProxy setup or default, but that any one can be changed directly from the screen using stick inputs....the param in each slot can be adjusted for it value also by stick input....sure you can figure out some wording ....

**Review by Hwurzburg — CHANGES_REQUESTED** [review]
I need to get a setup together and play with it some to make further comments....will try to do that by Monday PM

File: common/source/docs/common-paramosd.rst [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+:code:`MIN` allows to set a lower limit to the setteable range
+
+:code:`MAX` allows to limit the settable range's maximum
+
+:code:`INCREMENT` allows to use specific increments if required
+
+Examples
+--------
+
+To set OSD5 parameter1 to TRIM_THROTTLE, type :code:`osd param-show 5 1 TRIM_THROTTLE`
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:120]: param-show?

**Comment by Hwurzburg** [comment]
okay....loaded up master....started SITL with osd....changed OSD5_ENABLE=1 and set it to be displayed in disarm.....nothing but a blank OSD....tried to set some with MAVProxy (I can see there are defaults for page 5 and 6, maybe should have screen shots of those)....MAVProxy says the OSD commands are unknown commands...must have to update MAVProxy...that should be stated and how to do so

File: common/source/docs/common-paramosd.rst [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+    :width: 450px
+
+This allows to set and adjust paramaters using the ardupilot maxchip-based onboard OSD and an RC transmitter's stick input. 
+There's two additional OSD screens available (OSD5 and OSD6) that can hold up to 9 parameters each. Parameter values can be 
+adjusted in disarmed state only. The RC transmitter's roll and pitch stick input is used to navigate through the screen, pick 
+a parameter and adjust the parameter value.
+
+
+
+Additionally there's a "save and reboot" button on each screen that allows to save the adjusted settings and reboot the
```
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst:19]: defaults to "reboot". auto-switches to "save" if some setting is altered, switches back to "reboot" after having saved. added to text body.

File: common/source/docs/common-paramosd.rst [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+Additionally there's a "save and reboot" button on each screen that allows to save the adjusted settings and reboot the 
+flightcontroller if required.
+
+Accessing parameter screens
+===========================
+
+To enable the parameter screens, set :ref:`OSD5_ENABLE <OSD5_ENABLE>` and :ref:`OSD6_ENABLE <OSD6_ENABLE>` = 1.
+Paramater screens can be accessed in the same way the conventional osd screens are accessed. 
+See :ref:`screens-and-screen-switching <common-osd-overview.html#screens-and-screen-switching>` for reference.
```
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed,

File: common/source/docs/common-paramosd.rst [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+.. image:: ../../../images/stickinput.jpg
+    :target: ../_images/stickinput.jpg
+    :width: 450px
+
+  
+Parameter slot scroll-selection
+===============================
+
+Holding roll right for > 3 seconds on a parameter's value will highlight both the selected parameter and its value (blink). This allows 
+to scroll through the full list of parameters and place another paramater into that respective slot using pitch up and down stick input.
```
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    :target: ../_images/stickinput.jpg
+    :width: 450px
+
+  
+Parameter slot scroll-selection
+===============================
+
+Holding roll right for > 3 seconds on a parameter's value will highlight both the selected parameter and its value (blink). This allows 
+to scroll through the full list of parameters and place another paramater into that respective slot using pitch up and down stick input. 
+Move left to exit parameter selection and got back to standard screen navigation.
```
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed.

File: common/source/docs/common-paramosd.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: most of the confusion likely added by my not exactly edifying (and slightly vinous) attempt to put that into some generally intelligible wording, but in a nutshell - yes, min / max / incr are best guesses based on metatdata, plus incr will adaptively auto-adjust to current value.
additionally, you can set min / max / incr to individual preferences, but the default guesses usually work pretty well.

i reworked that passage accordingly.

File: common/source/docs/common-paramosd.rst [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+To set OSD6 parameter2 to FLTMODE1 using respective mnemonics, type :code:`osd param-set 6 2 FLTMODE1 FLIGHT_MODE`
+
+To set OSD5 parameter2 to LIM_ROLL_CD, limiting range from 2000 - 6000 with increments of 100, type :code:`osd param-set 5 2 LIM_ROLL_CD 2000 6000 100`
+
+To show what OSD6 parameter2 is set to, type :code:`osd param-show 6 2`
+
+
+.. note::
+   
+   Mind that some :code:`TYPE` mnemonic options are vehicle-specific. Make sure to apply mnemonics to the matching vehicle type only to avoid unexpected results.
```
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 1 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 2 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 3 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 4 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed. no need to mention the chipset at all.

File: common/source/docs/common-paramosd.rst (hunk 5 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 6 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 8 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed.
i think the defaults andy chose are copter-only params (ATC_RAT_...). might make sense to add defaults for plane too.

File: common/source/docs/common-paramosd.rst (hunk 9 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst:131]: yes, this is correct.
we might talk to @andyp1per about the syntax. it's not super well aligned with common param handling syntax and type commands are case-sensitive, while non-osd param commands are not... but then, imho the final goal is to add GUI-type setting support in MP

**Comment by vierfuffzig** [comment]
thanks for your review @Hwurzburg ! 



> MAVProxy says the OSD commands are unknown commands

did you `module load osd` ?

**Comment by andyp1per** [comment]
@Hwurzburg I've not tried the displayed in disarm thing. I usually set OSD_CHAN=7 and then rc 7 1800 to switch to the screen in SITL. I'll try your disarm suggestion.

File: common/source/docs/common-paramosd.rst (hunk 9 above)
Comment (andyp1per) [reply common/source/docs/common-paramosd.rst:131]: I did want to fix the case sensitivity thing.
I'm open to other syntax suggestions, although what we have is somewhat consistent with parameter syntax. Note that we can't have:
```osd param show ....```
Or at least its a pain to implement, I would prefer just a single command name, and as @vierfuffzig ideally we want MP to implement this.

File: common/source/docs/common-paramosd.rst (hunk 8 above)
Comment (andyp1per) [reply common/source/docs/common-paramosd.rst]: Yes, you are right. Tell me what makes sense  here. I think also we should have some decent defaults for OSD6 that are different to OSD5

File: common/source/docs/common-paramosd.rst (hunk 9 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst:131]: @andyp1per just brainstorming disregarding the pita to implement;
- rename module to `paramosd` as `osd` is somewhat unspecific regarding the various osd options we have now
- make syntax `paramosd set` and `paramosd show` to align with module `param`

**Comment by vierfuffzig** [comment]
doesn't matter how you switch to `osd5`, it's just one of the default params being present in plane

**Comment by andyp1per** [comment]
Ahh got you. Ok so what should these be? My feeling was that one panel should be a tuning panel and one a config panel

**Comment by vierfuffzig** [comment]
i'm trying to find plane presets right now. plane isn't that much about tuning as copter is, so i'd suggest using OSD5 for config stuff and OSD6 for tuning et altera.

File: common/source/docs/common-paramosd.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst:19]: probably needs that explained in this...difference between saved and after a reboot.

**Comment by Hwurzburg** [comment]
here are my suggestions for the plane defaults...from using this ability in Yaapu's bi-dir FRSKY for many months bringing up at least 5 new planes and quadplanes:
Configuration
- ARMING_CHECK
- TRIM_THROTTLE
- TRIM_ARSPD_CM
- ARSPD_FBW_MIN
- ARSPD_FBW_MAX
- ALT_HOLD_RTL
- AHRS_TRIM_Y
- THR_MAX
- Q_TRIM_PITCH

TUNING

- RLL2SRV_P
- RLL2SRV_I
- RLL2SRV_D
- PTCH2SRV_P
- PTCH2SRV_I
- PTCH2SRV_D
- PTCH2SRV_RLL
- TUNE_PARAM
- TUNE_RANGE

**Comment by Hwurzburg** [comment]
tried again with copter sim off master sim_vehicle.py -v ArduCopter --osd --console ....OSD appears...changed OSD5_ENABLE =1 and OSD_DISARM_SCR to 5, rebooted....the OSD screen did not change from 1...
no... module load osd in MAVPRoxy gives module not found....as I said, I probably am using the MAVProxy everyone else is as loaded when the build environ is setup...if this is in a new release, then instructions must be added to this page on how to get the most recet version...but irrespective of MAVProxy version, shouldnt the screen switch to the panel 5 and show the params in disarm??

**Comment by vierfuffzig** [comment]
added note on MAVProxy version and update command.
@Hwurzburg i'm under the impression that OSD_ARM_SCR and _DSARM_SCR don't work at all, irrespective of param osd.

your plane default suggestions are very similar to what i use, except for ARMING_CHECK. i personally am absolutely not a fan of making it easier to mess with arming checks. whoever knows what he's doing will find it.

adding defaults per-screen plus vehicle-specific might not be that easy, anyways here's my suggestion for a single plane screen:

```
AP_OSD_ParamSetting params[NUM_PARAMS] = {
    //  {slot, en, x, y, key, idx , grp}
        {1, true, 2, 2, 232, 0, 1 },     // RLL2SRV_P
        {2, true, 2, 3, 233, 0, 1 },     // PTCH2SRV_P
        {3, true, 2, 4, 4, 0, 1094 },    // SERVO_AUTO_TRIM
        {4, true, 2, 5, 184, 0, 0 },     // TRIM_THROTTLE
        {5, true, 2, 6, 182, 0, 0 },     // THR_MAX
        {6, true, 2, 7, 189, 0, 0  },    // THR_SLEWRATE
        {7, true, 2, 8, 224, 0, 0 },     // WP_LOITER_RAD
        {8, true, 2, 9, 156, 0, 0 },     // ALT_HOLD_RTL
        {9, true, 2, 10, 9, 0, 9 }       // OSD_FONT
    };
```

**Comment by andyp1per** [comment]
@Hwurzburg you will need to reinstall mavproxy, tridge pushed out a new version just recently containing the changes.

Try pip install --upgrade--force-reinstall mavproxy

File: common/source/docs/common-paramosd.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst]: no you misunderstand...."..#screens-and...." is not allowed...you must go to the common-osd-overview doc and create a tag
-screens-and-switching above the title for that section...then use that in the <>

File: common/source/docs/common-paramosd.rst [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+   OSD based parameter adjustment is designed to allow basic adjustments without any aditional data link or GCS setup required. However, 
+   some parameters with exceptionally wide value ranges, as well as complex bitmap-type parameters might not be ideal for adjustment using parameter OSD.
+
+
+Parameter range, increment and type-mnemonics
+=============================================
+
+OSD based parameter menu uses parameter metadata to estimate a useful value range and incrementation for each parameter. Additionally, 
+this supports to display human-readable abbreviations for:
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:68]: suggestion:
Additionally, support for human-readable abbreviations is provided for:

File: common/source/docs/common-paramosd.rst [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+
+- serial protocols
+
+- RC aux options
+
+- failsave actions
+
+.. note::
+
+   This feature my have limitations on certain parameters with exceptionally wide value ranges and when handling bitmap-type parameters.
```
Comment (Hwurzburg) [inline common/source/docs/common-paramosd.rst:82]: may have

File: common/source/docs/common-paramosd.rst (hunk 8 above)
Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst]: perhaps one for plane and one for copter?

**Comment by Hwurzburg** [comment]
>..anyways here's my suggestion for a single plane screen:
> AP_OSD_ParamSetting params[NUM_PARAMS] = {
>     //  {slot, en, x, y, key, idx , grp}
>         {1, true, 2, 2, 232, 0, 1 },     // RLL2SRV_P
>         {2, true, 2, 3, 233, 0, 1 },     // PTCH2SRV_P
>         {3, true, 2, 4, 4, 0, 1094 },    // SERVO_AUTO_TRIM
>         {4, true, 2, 5, 184, 0, 0 },     // TRIM_THROTTLE
>         {5, true, 2, 6, 182, 0, 0 },     // THR_MAX
>         {6, true, 2, 7, 189, 0, 0  },    // THR_SLEWRATE
>         {7, true, 2, 8, 224, 0, 0 },     // WP_LOITER_RAD
>         {8, true, 2, 9, 156, 0, 0 },     // ALT_HOLD_RTL
>         {9, true, 2, 10, 9, 0, 9 }       // OSD_FONT
I agree with pid P's..adjusting TECs speeds is pretty critical in my view...ARSPD_FBW_MIN, TRIM_THROTTLE,TRIM_ARSPD_CM....THR_MAX....thats 6
why SERVO_AUTO_TRIM? it should be on all the time...
less passionate about the others...

**Comment by andyp1per** [comment]
https://github.com/ArduPilot/ardupilot/pull/15304 is nearly there - still some kinks to work out and need to zero in on the final set of parameters for both copter and plane

**Comment by vierfuffzig** [comment]
@Hwurzburg and @andyp1per thanks for review and respective updates, will fix documentation respectively.

on a sidenote i think ARM_SCR logic as is now used non-param screens only:
https://github.com/ArduPilot/ardupilot/blob/f78a09ba2a93a1792ac03e22babba54e6458b6f3/libraries/AP_OSD/AP_OSD.cpp#L337

File: common/source/docs/common-paramosd.rst (hunk 16 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

File: common/source/docs/common-paramosd.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: sry, got it now. will add that, simplified fix for now

**Comment by andyp1per** [comment]
Ok https://github.com/ArduPilot/ardupilot/pull/15304 is working now - be grateful if you could try it out. I combined both your suggestions, but I am not a plane user - so please make further suggestions.

The disarm screen thing is easy to fix, but once you are in the parameter screen you are prevented from arming, so does not work quite the same way. My suggestion is to leave the status quo but open to other ideas.

**Comment by Hwurzburg** [comment]
Basti, I added a tag to the OSD page for screen switching, if yo wish to use it:  screen-switching

**Comment by vierfuffzig** [comment]
thanks henry, back from family business, will push asap respectively. 
re type mnemonics: they translate parameter values (numbers) into human readable abbreviations, so that flightmodes, serial protocols or aux functions etc have names, not numbers, and you dont need to look up what exact flightmode „13“ translates to, or what number rc option „soar“ has.

File: common/source/docs/common-paramosd.rst (hunk 15 above)
Comment (Hwurzburg) [reply common/source/docs/common-paramosd.rst]: Revised suggestion:
Additionaly, support is provided to be able to display some parameter values in more readable form by the use of some mnemonic sets for certain parameters:

missed spelling on failsafe in that list

**Comment by vierfuffzig** [comment]
@andyp1per tested, working as expected. i'm happy with the defaults, they should be a good place to start from.

not being a python pro, will `pip install --upgrade pymavlink MAVProxy --user` work or is adding `--force-reinstall` mandatory?

File: common/source/docs/common-paramosd.rst (hunk 15 above)
Comment (vierfuffzig) [reply common/source/docs/common-paramosd.rst]: fixed

**Comment by Hwurzburg** [comment]
I am going to go ahead and merge this and link from the upcoming features page....if things change on the param setup side, we can update...thanks..!

**Comment by andyp1per** [comment]
The params were updated on Wednesday - we should probably redo the example screens

**Comment by vierfuffzig** [comment]
https://github.com/ArduPilot/ardupilot_wiki/pull/3036

**Pull request merged**